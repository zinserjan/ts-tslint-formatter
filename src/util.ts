import * as path from "path";
import { RuleFailure } from "tslint";
import normalizePath = require("normalize-path");
import Es6Error = require("es6-error");

export interface GroupedRuleFailures {
  [fileName: string]: Array<RuleFailure>;
}

export interface Formatter {
  (ruleFailures: Array<RuleFailure>, context: string): FormattedRuleFailure;
}

export class FormattedRuleFailure extends Es6Error {
  file?: string;

  constructor(message: string, file: string) {
    super(message);
    this.file = file;
  }
}

export const groupByFile = (ruleFailures: Array<RuleFailure>): GroupedRuleFailures => {
  return ruleFailures.reduce((groups: GroupedRuleFailures, ruleFailure: RuleFailure) => {
    const fileName = ruleFailure.getFileName();
    if (!groups[fileName]) {
      groups[fileName] = [];
    }
    groups[fileName].push(ruleFailure);
    return groups;
  }, {} as GroupedRuleFailures);
};

export const formatFilePath = (filePath: string, context: string) => {
  let fileName = normalizePath(path.relative(context, filePath));
  if (fileName && fileName[0] !== ".") {
    fileName = `./${fileName}`;
  }
  return fileName;
};
