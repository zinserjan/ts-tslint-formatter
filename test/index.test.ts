import * as os from "os";
import * as path from "path";
import { getTsConfigs, getRuleFailures } from "./util/testHelper";
import format from "../src/index";
const chalk = require("chalk");

const formatters = ["stylish"];
const tsconfigs = getTsConfigs();

formatters.forEach(formatter => {
  describe(`${formatter} formatter`, () => {
    tsconfigs.forEach((tsConfigPath: string) => {
      const dir = path.basename(path.dirname(tsConfigPath));
      const tslintConfig = path.join(path.dirname(tsConfigPath), "tslint.json");

      it(dir, () => {
        const failures = getRuleFailures(tsConfigPath, tslintConfig);
        const formattedFailures = format(failures, formatter);

        formattedFailures.forEach(formatted => {
          formatted.message = formatted.message.replace(os.EOL, "\n");
          expect({ file: formatted.file, message: formatted.message }).toMatchSnapshot();
        });
      });
    });
  });
});
