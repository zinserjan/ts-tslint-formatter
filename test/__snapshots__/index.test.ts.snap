// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stylish formatter error-1 1`] = `
Object {
  "file": "./test/fixture/error-1/src/entry.ts",
  "message": "  [2m3:19[22m  [31merror[39m  ' should be \\"  [2mquotemark[22m
",
}
`;

exports[`stylish formatter error-2 1`] = `
Object {
  "file": "./test/fixture/error-2/src/entry.ts",
  "message": "  [2m2:19[22m  [31merror[39m  ' should be \\"  [2mquotemark[22m
  [2m4:19[22m  [31merror[39m  ' should be \\"  [2mquotemark[22m
",
}
`;

exports[`stylish formatter warning-1 1`] = `
Object {
  "file": "./test/fixture/warning-1/src/entry.ts",
  "message": "  [2m3:19[22m  [33mwarning[39m  ' should be \\"  [2mquotemark[22m
",
}
`;

exports[`stylish formatter warning-2 1`] = `
Object {
  "file": "./test/fixture/warning-2/src/entry.ts",
  "message": "  [2m2:19[22m  [33mwarning[39m  ' should be \\"  [2mquotemark[22m
  [2m4:19[22m  [33mwarning[39m  ' should be \\"  [2mquotemark[22m
",
}
`;
